### PR TITLE
`gangway`: add status change endpoint

### PR DESCRIPTION
**Pull request by Prucek**
Resolves: https://github.com/kubernetes-sigs/prow/issues/115

Add new `BulkJobStatusChange` endpoint to change the status of multiple PJs.

### CODE REVIEW

File: pkg/gangway/gangway.go [hunk 1]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+
+	err, md := getHttpRequestHeaders(ctx)
+	if err != nil {
+		logrus.WithError(err).Debug("could not find request HTTP headers")
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	if err := request.Validate(); err != nil {
+		logrus.WithError(err).Debug("could not validate request fields")
+		return nil, err
```
Comment (smg247) [inline pkg/gangway/gangway.go:181]:
```suggestion
                return nil, status.Error(codes.InvalidArgument, err.Error())
```

File: pkg/gangway/gangway.go [hunk 2]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+	}
+
+	mainConfig := gw.ConfigAgent.Config()
+	allowedApiClient, err := mainConfig.IdentifyAllowedClient(md)
+	if err != nil {
+		logrus.WithError(err).Debug("could not find client in allowlist")
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	pjList, err := gw.ProwJobClient.List(ctx, metav1.ListOptions{})
```
Comment (smg247) [inline pkg/gangway/gangway.go:191]: We should put a TODO in here explaining why all prowjobs must be listed and how to fix it when it is possible

File: pkg/gangway/gangway.go [hunk 3]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+	jobExecutions := &JobExecutions{}
+	var count int32
+	for _, pj := range pjList.Items {
+		if !isMatchingCondition(pj, request) {
+			continue
+		}
+		if allowedApiClient != nil {
+			authorized := ClientAuthorized(allowedApiClient, pj)
+			if !authorized {
+				logrus.Error("client is not authorized to execute the given job")
```
Comment (smg247) [inline pkg/gangway/gangway.go:206]:
```suggestion
				logrus.Error("client is not authorized to modify the given job")
```

File: pkg/gangway/gangway.go [hunk 4]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+				return nil, status.Error(codes.PermissionDenied, "client is not authorized to execute the given job")
```
Comment (smg247) [inline pkg/gangway/gangway.go:207]:
```suggestion
				return nil, status.Error(codes.PermissionDenied, "client is not authorized to modify the given job")
```

File: pkg/gangway/gangway.go [hunk 5]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+			}
+		}
+
+		pj.Status.State = prowcrd.ProwJobState(strings.ToLower(request.GetJobStatusChange().GetDesired().String()))
+		updatedPj, err := gw.ProwJobClient.Update(ctx, &pj, metav1.UpdateOptions{})
+		if err != nil {
+			logrus.WithError(err).Debug("failed to update ProwJob status")
```
Comment (smg247) [inline pkg/gangway/gangway.go:214]: This seems like a bigger deal than a debug log. It at least needs to be a warning, but probably better off as an error level.

File: pkg/gangway/gangway.go [hunk 6]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+			continue
+		}
+		logrus.WithField("name", pj.Name).Info("ProwJob status updated to:", updatedPj.Status.State)
```
Comment (smg247) [inline pkg/gangway/gangway.go:217]:
```suggestion
		logrus.WithField("name", pj.Name).Infof("ProwJob status updated to: %s", updatedPj.Status.State)
```

File: pkg/gangway/gangway.go [hunk 7]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+
+		jobExec := &JobExecution{
+			Id:        updatedPj.Name,
+			JobName:   updatedPj.Spec.Job,
+			JobType:   JobExecutionType(TranslateProwJobType(updatedPj.Spec.Type)),
+			JobStatus: TranslateProwJobStatus(&updatedPj.Status),
+		}
+		jobExecutions.JobExecution = append(jobExecutions.JobExecution, jobExec)
+		count++
```
Comment (smg247) [inline pkg/gangway/gangway.go:226]: I don't think we need to keep track of the count like this, we can just do a `len(jobExecutions.JobExecution)` at the end, right?

File: pkg/gangway/gangway.go [hunk 8]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+	}
+
+	// If more than n jobs were affected, we only return the first n jobs
```
Comment (smg247) [inline pkg/gangway/gangway.go:229]: May as well use `10` rather than `n` in the comment

File: pkg/gangway/gangway.go [hunk 9]
```diff
@@ -135,13 +146,143 @@ func (gw *Gangway) GetJobExecution(ctx context.Context, gjer *GetJobExecutionReq
...
+	// because we don't want to return a huge list of jobs and wait for a long time
+	n := int32(10)
```
Comment (smg247) [inline pkg/gangway/gangway.go:231]: this should be defined as a const

File: pkg/gangway/gangway.go [hunk 10]
```diff
@@ -349,6 +490,17 @@ func (cjer *CreateJobExecutionRequest) Validate() error {
 	return nil
 }
 
+func (bjscr *BulkJobStatusChangeRequest) Validate() error {
```
Comment (smg247) [inline pkg/gangway/gangway.go:493]: Do we need to validate that the startedBefore and startedAfter are parseable as times if they are provided?

**Comment by petr-muller** [comment]
:+1: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_prow/155/pull-prow-integration/1791473589053034496#1:build-log.txt%3A887:

```
 === FAIL: test/integration/test TestGangwayBulkJobStatusChange (12.74s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1b69d80] 
```

Let's revert

**Pull request merged**